service/matching/user_data_manager.go

```diff
@@ -366,6 +382,49 @@ func (m *userDataManagerImpl) loadUserDataFromDB(ctx context.Context) error {
 	return nil
 }
 
+// Checks if data in db has not been modified since we loaded/modified it.
+func (m *userDataManagerImpl) refreshUserDataFromDB(ctx context.Context) error {
+	// Lock here to ensure we're not in the middle of an update, otherwise we may incorrectly
+	// think the db has old data if we update between read and verify.
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	response, err := m.store.GetTaskQueueUserData(ctx, &persistence.GetTaskQueueUserDataRequest{
+		NamespaceID: m.partition.NamespaceId(),
+		TaskQueue:   m.partition.TaskQueue().Name(),
+	})
+	if common.IsNotFoundError(err) {
+		response, err = &persistence.GetTaskQueueUserDataResponse{}, nil
+	}
+	if err != nil {
+		return err
+	}
+
+	if response.UserData.GetVersion() == m.userData.GetVersion() {
+		return nil
+	}
+
+	tags := []tag.Tag{
+		tag.UserDataVersion(response.UserData.GetVersion()),
+		tag.NewInt64("expected-user-data-version", m.userData.GetVersion()),
+		tag.Timestamp(hybrid_logical_clock.UTC(response.UserData.GetData().GetClock())),
+		tag.NewTimeTag("expected-user-data-timestamp", hybrid_logical_clock.UTC(m.userData.GetData().GetClock())),
+	}
+
+	if response.UserData.GetVersion() < m.userData.GetVersion() {
+		// We have newer data in memory than the db. This should only happen if the database
+		// went back in time. We should unload and start over.
```

Comment:
just to confirm: you're okay with the behavior of unloading if we see older data, and of loading new data if we see new data?